tests/storage/test_registration.py

```diff
@@ -166,3 +168,101 @@ def test_3pid_inhibit_invalid_validation_session_error(self) -> None:
             ThreepidValidationError,
         )
         self.assertEqual(e.value.msg, "Validation token not found or has expired", e)
+
+
+class ApprovalRequiredRegistrationTestCase(HomeserverTestCase):
+    def default_config(self) -> JsonDict:
+        config = super().default_config()
+
+        # If there's already some config for this feature in the default config, it
+        # means we're overriding it with @override_config. In this case we don't want
+        # to do anything more with it.
+        msc3866_config = config.get("experimental_features", {}).get("msc3866")
+        if msc3866_config is not None:
+            return config
+
+        # Require approval for all new accounts.
+        config["experimental_features"] = {
+            "msc3866": {
+                "enabled": True,
+                "require_approval_for_new_accounts": True,
+            }
+        }
+        return config
+
+    def prepare(self, reactor: MemoryReactor, clock: Clock, hs: HomeServer) -> None:
+        self.store = hs.get_datastores().main
+        self.user_id = "@my-user:test"
+        self.pwhash = "{xx1}123456789"
+
+    @override_config(
+        {
+            "experimental_features": {
+                "msc3866": {
+                    "enabled": True,
+                    "require_approval_for_new_accounts": False,
+                }
+            }
+        }
+    )
+    def test_approval_not_required(self) -> None:
+        """Tests that if we don't require approval for new accounts, newly created
+        accounts are automatically marked as approved.
+        """
+        self.get_success(self.store.register_user(self.user_id, self.pwhash))
+
+        user = self.get_success(self.store.get_user_by_id(self.user_id))
```

Comment:
since `get_user_by_id` is deprecated, please could you update this to use `get_userinfo_by_id` instead?

At the same time, you can do the `None → True` conversion whilst loading it into the `UserInfo` struct.

Reply:
I'd rather test the function that's being modified directly, rather than doing this indirectly by testing a function that calls it. Plus there are still loads of places calling `get_user_by_id` and not `get_userinfo_by_id` in the code.

Reply:
just feels like we're propagating the use of a deprecated function here rather than its light wrapper which is the nominated successor. But OK, all the more reason to remove it soon